Config as code strategies (#705)
OctopusDeploy/blog: blog/2022-03/config-as-code-strategies/index.md

---
title: Config as Code strategies
description: TBC
author: steve.fenton@octopus.com
visibility: private
published: 9999-01-01
metaImage: blogimage-config-as-code.png
bannerImage: blogimage-config-as-code.png
bannerImageAlt: Configuration as code
isFeatured: false
tags:
 - DevOps
 - Configuration as Code 
---

> Since last year, when we released the early-access preview of Octopus Config as Code, there have been many questions about how to use the feature to get the best results. This article will explain some good practices for using config as code and how to adjust your strategy in different situations.

Reviewer: Slight tweak to the opening sentence:
```suggestion
Since we released the early-access preview of Octopus Config as Code last year, there have been many questions about how to use the feature for the best results. This article explains some good practices for using config as code and how to adjust your strategy in different situations.
```

## Why use config as code?

Git is the perfect solution for versioning code and keeping track of changes over time. It has established patterns for branching the code and for publishing and approving changes. It also allows you to compare versions and travel back in time if you need to.

> The Octopus Config as Code feature enables you to store your deployment process as configuration files in a Git repository rather than in the Octopus database. You can use config as code to:

Reviewer: Enables always feels like a complicated word than is necessary :)
```suggestion
The Octopus Config as Code feature lets you store your deployment process as configuration files in a Git repository rather than in the Octopus database. You can use config as code to:
```

The configuration is stored as human-readable files that use Octopus Configuration Language (OCL). OCL was designed to make it easier to read and edit the deployment process and review any changes. There is a [Visual Studio Code extension to make it easier to work with OCL files](https://marketplace.visualstudio.com/items?itemName=octopusdeploy.vscode-octopusdeploy).

Not everything is moved into the repository when you enable version control. A list of version-controlled resources is available in our [configuration as code reference](https://octopus.com/docs/projects/version-control/config-as-code-reference).

:::warning
Switching on version control for your project is a one-way change. You can't move the project back into the Octopus database once it's in a repository. You can clone an existing project to try config as code and confirm that it meets your needs before enabling it for your production projects.
:::

## Where to store your configuration

One of the first decisions you will need to make is where to store your deployment configuration files. You can keep your configuration:

- Alongside your application code
- In a separate deployment repository

> Each option is described below to explain when they work and when to avoid them. You might have noticed that you can arrange these possibilities along a scale from a one-to-one relationship with applications to a single large repository. We recommend keeping your deployment configuration in the same repository as the application code, but there are specific circumstances where the other options may be suitable.

Reviewer: ```suggestion
Read on, to learn when each option works best and when to avoid them. You might have noticed that you can arrange these possibilities along a scale from a one-to-one relationship with applications to a single large repository. We recommend keeping your deployment configuration in the same repository as the application code, but there are specific circumstances where the other options may be suitable.
```

After setting up config as code, you can [move your deployment configuration files](https://octopus.com/docs/projects/version-control/moving-version-control) if you change your mind.

### Alongside application code

Placing your deployment configuration alongside your application code is the pattern we recommend. It is best  _to evolve your deployment process alongside your application code_. Putting the configuration in the same location as the application aligns with DevOps practices, where engineers take end-to-end responsibility for their applications.

If you opt to store your configuration in the application repository, each application will have its own `.octopus` directory with the configuration files. This arrangement makes it easy to find the appropriate deployment process for each application.

> If you don't want changes to the deployment process to trigger a build, you can mask the `.octopus` folder in your build server.

Reviewer: It might not be necessary, but I wonder if it's worth expanding on this point a bit just to make it really clear what's been built?
```suggestion
If you don't want changes to the deployment process to trigger a build of your application, you can mask the `.octopus` folder in your build server.
```
